File: Design-Patterns-Using-Cpp/ArchitecturalPatterns/MVP.cpp
#include <iostream>
#include <string>

class Model {
private:
    std::string data;

public:
    Model() : data("Hello") {}

    void setData(const std::string& data) {
        std::cout << "Model: Set data: " << data << std::endl;
        this->data = data;
    }

    std::string getData() {
        std::cout << "Model: Get data: " << this->data << std::endl;
        return this->data;
    }
};

class View {
public:
    std::string getUserInput() {
        std::cout << "View: getUserInput" << std::endl;
        std::cout << "View: Enter user input: ";
        std::string userInput = "Hello, World!";
        std::cout << userInput << std::endl;
        // std::getline(std::cin, userInput); // Uncomment for user input
        return userInput;
    }

    void displayData(const std::string& data) {
        std::cout << "View: Display Result: " << data << std::endl;
    }
};

class Presenter {
private:
    Model* model;
    View* view;

public:
    Presenter(Model* model, View* view) : model(model), view(view) {}

    void onButtonClick() {
        std::cout << "Presenter: onButtonClick." << std::endl;
        std::string data = this->view->getUserInput();
        this->model->setData(data);
        data = this->model->getData();
        this->view->displayData(data);
    }
};

// Client code
int main() {
    Model model;
    View view;
    Presenter presenter(&model, &view);
    presenter.onButtonClick();
    return 0;
}

/*
Presenter: onButtonClick.
View: getUserInput
View: Enter user input: Hello, World!
Model: Set data: Hello, World!
Model: Get data: Hello, World!
View: Display Result: Hello, World!
*/

File: Design-Patterns-Using-Go/ArchitecturalPatterns/MVP.go
package main

import (
	"fmt"
	"bufio"
	"os"
)

// Model struct
type Model struct {
	data string
}

// NewModel constructor
func NewModel() *Model {
	return &Model{
		data: "Hello",
	}
}

// SetData method
func (m *Model) SetData(data string) {
	fmt.Println("Model: Set data:", data)
	m.data = data
}

// GetData method
func (m *Model) GetData() string {
	fmt.Println("Model: Get data:", m.data)
	return m.data
}

// View struct
type View struct{}

// GetUserInput method
func (v *View) GetUserInput() string {
	fmt.Println("View: getUserInput")
	//fmt.Println("View: Enter user input: Hello, World!")
	//return "Hello, World!"
    fmt.Print("View: Enter user input: ")
    scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	userInput := scanner.Text()
	return userInput
}

// DisplayData method
func (v *View) DisplayData(data string) {
	fmt.Println("View: Display Result:", data)
}

// Presenter struct
type Presenter struct {
	model *Model
	view  *View
}

// NewPresenter constructor
func NewPresenter(model *Model, view *View) *Presenter {
	return &Presenter{
		model: model,
		view:  view,
	}
}

// OnButtonClick method
func (p *Presenter) OnButtonClick() {
	fmt.Println("Presenter: onButtonClick.")
	data := p.view.GetUserInput()
	p.model.SetData(data)
	data = p.model.GetData()
	p.view.DisplayData(data)
}

// Client Code
func main() {
	model := NewModel()
	view := &View{}
	presenter := NewPresenter(model, view)
	presenter.OnButtonClick()
}

/*
Presenter: onButtonClick.
View: getUserInput
View: Enter user input: Hello, World!
Model: Set data: Hello, World!
Model: Get data: Hello, World!
View: Display Result: Hello, World!
*/